Billing unit tests

milo-hyben opened the pull request:
This PR contains:
1. BqTest class introduced to allow for BigQuery mocking.
2. Unit tests for all the billing python code, bringing the total metamist code coverage to 76%.
3. Members functions in billing were converted to static method where possible.
4. A few variables were renamed

illusional commented on test/test_bq_billing_base.py:
```diff
@@ -0,0 +1,885 @@
+
+        func_filter = BillingBaseTable._prepare_labels_function(query)
+
+        self.assertEqual('getLabelValue', func_filter.func_name)
```
Are these tests still relevant with the label changes recently?

milo-hyben replied:
Valid point, there are not used after we made change to the BQ table layout, 
let me remove all those labelFunction tests.

illusional reviewed (commented):
Broadly looking good, are there changes to the label bits that need to be made with the table change?

I dropped a few comments in the middle, but can you split out tests that are testing multiple things (unless it's really small like testing None + 1 value), but if it requires a new set of mocking, then IMO a new test it best.

I think it's worth catching and validating queries if the instance is available, mostly just adding some auth into the GitHub action.

milo-hyben commented on db/python/tables/bq/billing_filter.py:
```diff
@@ -46,3 +47,17 @@ class BillingFilter(GenericBQFilterModel):
+
+        # compare all attributes
+        for att in self.__dict__:
+            if not getattr(self, att).__eq__(getattr(other, att)):
```
Make sense.

milo-hyben commented on test/test_api_billing.py:
```diff
@@ -0,0 +1,280 @@
+TEST_API_BILLING_USER = 'test_user'
+
+
+class TestApiBilling(BqTest):
```
Good idea, make one default on the top.

milo-hyben commented on test/testbqbase.py:
```diff
@@ -0,0 +1,47 @@
+
+    # store connections here, so they can be created PER-CLASS
+    # and don't get recreated per test.
+
```
Removed

milo-hyben commented on test/test_api_utils.py:
```diff
@@ -0,0 +1,77 @@
+
+        # there is 3 (INVOICE_DAY_DIFF) days difference between invoice month st and end
+        self.assertEqual(
+            (datetime(2020, 12, 29).date(), datetime(2021, 2, 3).date()),
```
Yes, of course :-)

milo-hyben commented on test/test_bq_billing_base.py:
```diff
@@ -0,0 +1,885 @@
+    This returns one mockup BQ query result for various SQL queries
+    This is used in a few tests
+    """
+    if 'month.cost as monthly_cost' in query:
```
Refactored to two separate functions. 
One of the mockup function stills needs if statement as BQ is called twice in one Api call and you can only have one mockup function.

milo-hyben commented on test/test_bq_billing_base.py:
```diff
@@ -0,0 +1,885 @@
+        query_job_result.result.return_value = mock_rows
+
+        single_row = BillingBaseTable._convert_output(query_job_result)
+        self.assertEqual([{}], single_row)
```
Have split them as suggested.

milo-hyben commented on test/test_bq_billing_gcp_daily.py:
```diff
@@ -0,0 +1,194 @@
+            '        ON month.field = day.field\n'
+            '        AND month.cost_category = day.cost_category\n'
+            '        '
+        )
```
Very good point!

milo-hyben commented on test/test_bq_billing_base.py:
```diff
@@ -0,0 +1,885 @@
+    return []
+
+
+class TestBillingBaseTable(BqTest):
```
Regarding Cost of query I will create a new PR for that.
Having minimal BQ metadata similar is a good idea, however it start smelling as integration tests :-)
Any I will put it on the list.

milo-hyben commented:
Hi Michael, I had addresses most of your comment. The only left for the future PR is to use local BQ installation to run those tests against BQ tables. As well cost of query I will add to the response in a different PR.

illusional reviewed (approved):
😮‍💨

milo-hyben commented on test/test_api_billing.py:
```diff
@@ -14,6 +14,9 @@
+patch('api.routes.billing.is_billing_enabled', return_value=True)
+
 
 class TestApiBilling(BqTest):
```
I guess it is treated as a function call and not a decorator.

illusional replied:
Oh it's supposed to have a `@` in front of it!!

milo-hyben replied:
Nope, that can only be done in front of a function call. I have tried it did not work for the class.
